File: packages/dart/lib/src/keyid.dart
import 'dart:typed_data';

import 'base64url.dart';
import 'crypto_provider.dart';
import 'errors.dart';
import 'openpgp.dart';
import 'pkcs8.dart';
import 'types.dart';

const _hex = '0123456789ABCDEF';

Future<String> absoluteKeyId(
  List<int> canonicalPublicKey,
  CkvfCrypto crypto,
) async {
  final digest = await crypto.sha256(canonicalPublicKey);
  return bytesToBase64url(digest);
}

/// Short Key IDs are lookup hints and are not unique identifiers.
String shortKeyIdFromDigest(List<int> digest) {
  if (digest.length < 4) fail('ERR_SHORT_KEY_ID', 'digest too short');
  final hex = StringBuffer();
  for (var i = 0; i < 4; i++) {
    final b = digest[i];
    hex.write(_hex[(b >> 4) & 15]);
    hex.write(_hex[b & 15]);
  }
  final s = hex.toString();
  return '${s.substring(0, 4)}-${s.substring(4, 8)}';
}

Future<String> shortKeyId(List<int> canonicalPublicKey, CkvfCrypto crypto) async {
  final digest = await crypto.sha256(canonicalPublicKey);
  return shortKeyIdFromDigest(digest);
}

Future<({String absoluteKeyId, String shortKeyId})> keyIds(
  List<int> canonicalPublicKey,
  CkvfCrypto crypto,
) async {
  final digest = await crypto.sha256(canonicalPublicKey);
  return (
    absoluteKeyId: bytesToBase64url(digest),
    shortKeyId: shortKeyIdFromDigest(digest),
  );
}

Uint8List canonicalPublicKeyBytes(
  KeyEncoding encoding,
  String publicKeyB64, [
  String? privateKeyB64,
]) {
  final pub = base64urlToBytes(publicKeyB64);
  if (encoding == 'openpgp-tsk') {
    return canonicalOpenPgpPublicKey(
      pub,
      privateKeyB64 == null ? null : base64urlToBytes(privateKeyB64),
    );
  }
  if (encoding == 'pkcs8' || encoding == 'pkcs12') {
    if (_looksLikeSpki(pub)) return pub;
    if (privateKeyB64 != null) {
      try {
        return encoding == 'pkcs8'
            ? spkiFromPkcs8(base64urlToBytes(privateKeyB64))
            : pub;
      } catch (_) {
        return pub;
      }
    }
    return pub;
  }
  fail('ERR_ENCODING', 'unsupported encoding $encoding');
}

bool _looksLikeSpki(List<int> bytes) => bytes.length >= 2 && bytes[0] == 0x30;

File: packages/js/packages/core/src/keyid.ts
import { fail } from "./errors.js";
import { bytesToBase64url, base64urlToBytes } from "./base64url.js";
import type { CkvfCrypto } from "./crypto-provider.js";
import type { KeyEncoding } from "./types.js";
import { canonicalOpenPgpPublicKey } from "./openpgp.js";
import { spkiFromPkcs8 } from "./pkcs8.js";

const HEX = "0123456789ABCDEF";

export async function absoluteKeyId(canonicalPublicKey: Uint8Array, crypto: CkvfCrypto): Promise<string> {
  const digest = await crypto.sha256(canonicalPublicKey);
  return bytesToBase64url(digest);
}

/**
 * Short Key IDs are lookup hints and are not unique identifiers.
 * Applications MUST tolerate collisions. absolute_key_id is authoritative.
 */
export function shortKeyIdFromDigest(digest: Uint8Array): string {
  if (digest.length < 4) fail("ERR_SHORT_KEY_ID", "digest too short");
  const hex = [...digest.subarray(0, 4)].map((b) => HEX[(b >> 4) & 15]! + HEX[b & 15]!).join("");
  return `${hex.slice(0, 4)}-${hex.slice(4, 8)}`;
}

export async function shortKeyId(canonicalPublicKey: Uint8Array, crypto: CkvfCrypto): Promise<string> {
  const digest = await crypto.sha256(canonicalPublicKey);
  return shortKeyIdFromDigest(digest);
}

export async function keyIds(
  canonicalPublicKey: Uint8Array,
  crypto: CkvfCrypto,
): Promise<{ absolute_key_id: string; short_key_id: string }> {
  const digest = await crypto.sha256(canonicalPublicKey);
  return {
    absolute_key_id: bytesToBase64url(digest),
    short_key_id: shortKeyIdFromDigest(digest),
  };
}

export function canonicalPublicKeyBytes(encoding: KeyEncoding, publicKeyB64: string, privateKeyB64?: string | null): Uint8Array {
  const pub = base64urlToBytes(publicKeyB64);
  if (encoding === "openpgp-tsk") {
    return canonicalOpenPgpPublicKey(pub, privateKeyB64 ? base64urlToBytes(privateKeyB64) : undefined);
  }
  if (encoding === "pkcs8" || encoding === "pkcs12") {
    if (looksLikeSpki(pub)) return pub;
    if (privateKeyB64) {
      try {
        return encoding === "pkcs8" ? spkiFromPkcs8(base64urlToBytes(privateKeyB64)) : pub;
      } catch {
        return pub;
      }
    }
    return pub;
  }
  fail("ERR_ENCODING", `unsupported encoding ${encoding}`);
}

function looksLikeSpki(bytes: Uint8Array): boolean {
  return bytes.length >= 2 && bytes[0] === 0x30;
}
